### PR TITLE
[vlive] Add CH+ support (closes #16887)

### LINKED ISSUE #16887: How to login to vlive.tv to download paid contents?
### Make sure you are using the *latest* version: run `youtube-dl --version` and ensure your version is *2018.06.25*. If it's not, read [this FAQ entry](https://github.com/rg3/youtube-dl/blob/master/README.md#how-do-i-update-youtube-dl) and update. Issues with outdated version will be rejected.
- [x] I've **verified** and **I assure** that I'm running youtube-dl **2018.06.25**

### What is the purpose of your *issue*?
- [ ] Bug report (encountered problems with youtube-dl)
- [ ] Site support request (request for adding support for a new site)
- [ ] Feature request (request for a new functionality)
- [x] Question
- [ ] Other

I'm trying to download a video on vlive.tv. In vlive.tv, there are channel+, where only a paid user can view the contents, similar to youtube red(now premium). 

However, this sites supports multiple logins like Naver, Line, Facebook, and Twitter. I currently login via facebook. Just typing in my facebook email address, it doesn't work.

```
PS C:\Users\Administrator\Downloads> .\youtube-dl.exe -u myemail@gmail.com https://www.vlive.tv/video/66178
Type account password and press [Return]:
[vlive] 66178: Downloading webpage
ERROR: Coming soon!
PS C:\Users\Administrator\Downloads>
```

Any suggestions?

### PR DESCRIPTION
**Pull request by biwubo**
## Please follow the guide below

- You will be asked some questions, please read them **carefully** and answer honestly
- Put an `x` into all the boxes [ ] relevant to your *pull request* (like that [x])
- Use *Preview* tab to see how your *pull request* will actually look like

---

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [adding new extractor tutorial](https://github.com/ytdl-org/youtube-dl#adding-support-for-a-new-site) and [youtube-dl coding conventions](https://github.com/ytdl-org/youtube-dl#youtube-dl-coding-conventions) sections
- [x] [Searched](https://github.com/ytdl-org/youtube-dl/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into youtube-dl each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Bug fix
- [ ] Improvement
- [ ] New extractor
- [x] New feature

---

### Description of your *pull request* and other information

Adding support for CH+ videos that you can watch in the browser. I will ask one of the subscribers that shared their account credentials with me to also share them with @dstftw.

### CODE REVIEW

**Comment by biwubo** [comment]
@dstftw Hi, I believe you have been sent credentials by a CH+ subscriber now. The subscription on the account will have a limited validity so I would appreciate if you could review this PR before it expires. There appear to be a fair number of people anticipating CH+ support being added to youtube-dl.

**Comment by skullface** [comment]
Thanks for this!

**Comment by biwubo** [comment]
Hi @skullface, just wondering if you had tried using the changes in this PR.
If you have, could you let me know which channel(s) you used it for and if it worked? If you haven't then no worries, thanks for showing support!

The reason I ask is because the accounts I was given access to were all BTS subscribed accounts and I wanted to ensure it worked for other channels, too.

File: youtube_dl/extractor/vlive.py [hunk 1]
```diff
@@ -91,13 +136,15 @@ def _real_extract(self, url):
...
+        elif status == 'ONLY_APP':
+            raise ExtractorError('Unsupported video type', expected=True)
         else:
             raise ExtractorError('Unknown status %s' % status)
 
     def _get_common_fields(self, webpage):
         title = self._og_search_title(webpage)
         creator = self._html_search_regex(
-            r'<div[^>]+class="info_area"[^>]*>\s*<a\s+[^>]*>([^<]+)',
+            r'<div[^>]+class="info_area"[^>]*>\s*(?:<em[^>]*>.*</em\s*>\s*)?<a\s+[^>]*>([^<]+)',
```
Comment (dstftw) [inline youtube_dl/extractor/vlive.py]: Non greedy.

File: youtube_dl/extractor/vlive.py [hunk 2]
```diff
@@ -140,6 +180,16 @@ def _live(self, video_id, webpage):
         return info
 
     def _replay(self, video_id, webpage, long_video_id, key):
+        if '' in (long_video_id, key):
+            init_page = self._download_init_page(video_id)
+            video_info = self._parse_json(self._search_regex(
+                r'(?s)oVideoStatus\s*=\s*({.*})', init_page, 'video info'),
```
Comment (dstftw) [inline youtube_dl/extractor/vlive.py]: Don't capture empty dict.

File: youtube_dl/extractor/vlive.py [hunk 3]
```diff
@@ -47,12 +48,59 @@ class VLiveIE(InfoExtractor):
...
+        def is_logged_in():
+            login_info = self._download_json(
+                'https://www.vlive.tv/auth/loginInfo', None,
+                note='Downloading login info',
+                headers={'Referer': 'https://www.vlive.tv/home'})
+
+            return try_get(login_info,
+                           lambda x: x['message']['login'], bool) or False
+
+        if is_logged_in():
```
[the remark this answers is not in the document]
Comment (biwubo) [reply youtube_dl/extractor/vlive.py]: If you pass in cookies it can be.

[the remark this answers is not in the document]
Comment (dstftw) [reply youtube_dl/extractor/vlive.py]: If you pass cookie you must not pass credentials.

File: youtube_dl/extractor/vlive.py (hunk 1 above)
Comment (biwubo) [reply youtube_dl/extractor/vlive.py]: Just so I'm clear all you're asking is that I add another question mark after the non capturing group?

File: youtube_dl/extractor/vlive.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (biwubo) [reply youtube_dl/extractor/vlive.py]: Ah ok, I'll remove that then.

**Comment by biwubo** [comment]
Thanks a lot for the feedback and making the changes, I'll keep those in mind for next time.

**Pull request closed without merging**